### PR TITLE
reduce announcements of the automatic schema changes 

### LINKED ISSUE #15437: automatic schema modifications should be announced once
There are some schema modifications performed automatically (during bootstrap, upgrade etc.) by Scylla that are announced by multiple calls to `migration_manager::announce` even though they are logically one change. All of them should be refactored to announce only once. An example is the FIXME in `system_distributed_keyspace::start`:
```
// FIXME: fix this code to `announce` once
```
In #15435, we detected that we need to catch `group0_concurrent_modification` and retry `announce` in multiple places to fix the concurrent bootstrap in the Raft-based topology. However, before doing that, we should do the refactor mentioned in this issue to make "writing retry loops" easier.

### PR DESCRIPTION
**Pull request by patjed41**
There are some schema modifications performed automatically (during bootstrap, upgrade etc.) by Scylla that are announced by multiple calls to `migration_manager::announce` even though they are logically one change. Precisely, they appear in:
- `system_distributed_keyspace::start`,
- `redis:create_keyspace_if_not_exists_impl`,
- `table_helper::setup_keyspace` (for the `system_traces` keyspace).

All these places contain a FIXME telling us to `announce` only once. There are a few reasons for this:
- calling `migration_manager::announce` with Raft is quite expensive -- taking a `read_barrier` is necessary, and that requires contacting a leader, which then must contact a quorum,
- we must implement a retrying mechanism for every automatic `announce` if `group0_concurrent_modification` occurs to enable support for concurrent bootstrap in Raft-based topology. Doing it before the FIXMEs mentioned above would be harder, and fixing the FIXMEs later would also be harder.

This PR fixes the first two FIXMEs and improves the situation with the last one by reducing the number of the `announce` calls to two. Unfortunately, reducing this number to one requires a big refactor. We can do it as a follow-up to a new, more specific issue. Also, we leave a new FIXME.

Fixing the first two FIXMEs required enabling the announcement of a keyspace together with its tables. Until now, the code responsible for preparing mutations for a new table could assume the existence of the keyspace. This assumption wasn't necessary, but removing it required some refactoring.

Fixes #15437

### CODE REVIEW

**Comment by patjed41** [comment]
I opened #15595, which is a part of #15437 that this PR doesn't do. The new issue is more specific about the work left to do.

**Comment by kbr-scylla** [comment]
@nyh please review, you might be interested in this, as it's related to similar problems in alternator https://github.com/scylladb/scylladb/issues/13152

File: db/system_distributed_keyspace.cc
Comment (kbr-scylla) [reply db/system_distributed_keyspace.cc:309]: The test should be done both in raft and non-raft mode.

File: table_helper.cc
Comment (kbr-scylla) [inline table_helper.cc:149]: Is `setup_table` still used after this? If not, we could get rid of it.

File: db/system_distributed_keyspace.cc
Comment (patjed41) [reply db/system_distributed_keyspace.cc:237]: Are you sure it should be in a separate commit? Since removing the `announce` and leaving the `catch` doesn't make much sense, I think they should be done together (with an additional explanation in the commit message).

Comment (kbr-scylla) [reply db/system_distributed_keyspace.cc:237]: Easier to review in separate commit, because it's mostly an indentation change.

Alternatively, remove the `catch` in this commit, but leave the indentation as it was (i.e. "incorrect"), and in separate commit, fix the indentation.

In fact, in a bunch of commits I saw that some code was re-indented (because you removed a scope), which made the diff more complex to review. You could alter those commits so that indentation remains unchanged, and then have one large commit which fixes indentation throughout. The next reviewer after me would probably be grateful for that.

File: service/migration_manager.cc
Comment (patjed41) [reply service/migration_manager.cc]: TBH, I don't know what kind of mutation is included here.

However, we shouldn't call `include_keyspace` in the new `prepare_new_column_family_announcement` anyway. This function is called when the keyspace and the table are created together. We don't have to worry about a node missing a CREATE KEYSPACE migration and we can't assume that the keyspace exists.

I changed the code so that only the old `prepare_new_column_family_announcement` calls `include_keyspace` .

Comment (kbr-scylla) [reply service/migration_manager.cc]: > We don't have to worry about a node missing a CREATE KEYSPACE migration and we can't assume that the keyspace exists.

I'm not sure if `include_keyspace` actually assumes that the keyspace exists. I would guess that if the schema definition is not present on disk, it would just add an empty mutation that does nothing.

And I think we **do** have to worry about a potentially missing `CREATE KEYSPACE` -- the issue solved by `include_keyspace` is still potentially present in non-Raft mode, which also uses these `prepare_` functions. (In Raft mode, we indeed don't have to worry.)

File: db/system_distributed_keyspace.cc
Comment (patjed41) [reply db/system_distributed_keyspace.cc]: I applied the suggestions in the second and the third paragraphs. I added two commits that fix indentations -- 1 for `system_distributed_keyspace::start` and 1 for `create_keyspace_if_not_exists_impl`.

It seems like a really good practice and I'll try to follow this in the future. It's also good that you pointed it here, before I send the "retry `announce`s on concurrent operations" PR, which is all about changing indentations.

File: db/system_distributed_keyspace.cc
Comment (patjed41) [reply db/system_distributed_keyspace.cc:302]: I've tested it and confirm that this code works.

File: table_helper.cc
Comment (patjed41) [reply table_helper.cc:149]: In theory, `setup_table` can be used. It is used in `table_helper::cache_table_info` to handle a case when the table doesn't exist. Additionally, it is a public function that could be used in the future if `table_helper` found a new use case.

In practice, I don't think it's used. `table_helper` is used only for `system_traces` and its tables that are created during bootstrap. So, this case in `table_helper::cache_table_info` shouldn't happen.

I'm not sure what to do with this. `table_helper` requires a big refactor (or even removal) anyway as mentioned in #15595. Maybe we can leave `setup_table` for now and think about what to do with `table_helper` later.

File: service/migration_manager.cc
Comment (patjed41) [reply service/migration_manager.cc:701]: Thank you for the detailed explanation. I undid the removing of `include_keyspace`.

**Comment by patjed41** [comment]
v2:
- removed the `catch` expression and the FIXME from `get_updated_service_levels`
- moved indentation fixes to two new separate commits

File: service/tablet_allocator.cc
Comment (kbr-scylla) [inline service/tablet_allocator.cc:828]: @tgrabiec please review this patch (` tablet_allocator: update on_before_create_column_family `)

Comment (kbr-scylla) [reply service/tablet_allocator.cc:828]: Is this safe to do even in the case where keyspace already exists?

Comment (patjed41) [reply service/tablet_allocator.cc:828]: If it is unsafe, can we just check whether the keyspace exists and run the old code if it does? If we can, then the comment above `migration_listener::on_before_create_column_family`:
```
// ... Therefore, listeners can't load the keyspace from the database.
// Instead, they should use the `ksm` parameter if needed.
```
is misleading and I should change it.

Comment (kbr-scylla) [reply service/tablet_allocator.cc:828]: > If it is unsafe

I don't know that. You should investigate it.

The question is, in the case of keyspace already existing, whether this `create_replication_strategy` will return the same thing as `ks.get_replication_strategy()` would. If it turns out these replication strategies are different in a way that may affect the behavior of below code, this is unsafe.

Even if the keyspace does not exist yet, it could be that creating it first and then calling `get_replication_strategy()` would return a different thing than doing `create_replication_strategy` directly here, again causing different behavior of below code. (I don't believe that this happens though, after all, whatever `get_replication_strategy()` returns must have been created somehow, we just need to find a way to recreate the same thing. But you should check.)

**Review by piodul — COMMENTED** [review]
LGTM after the issue about calling `abstract_replication_strategy::create_replication_strategy` in `tablet_allocator::impl::on_before_create_column_family` is resolved

**Comment by tgrabiec** [comment]
I wonder if we really need to support this, it seems to add some ugly complexity (the create_keyspace_metadata thing). 

> calling migration_manager::announce with Raft is quite expensive -- taking a read_barrier is necessary, and that requires contacting a leader, which then must contact a quorum,

I'd say that extra group0 command on first boot or upgrade is irrelevant.

> we must implement a retrying mechanism for every automatic announce if group0_concurrent_modification occurs to enable support for concurrent bootstrap in Raft-based topology. Doing it before the FIXMEs mentioned above would be harder, and fixing the FIXMEs later would also be harder.

Can't we retry keyspace creation, and then retry table creation?

**Comment by nyh** [comment]
> Fixing the first two FIXMEs required enabling the announcement of a keyspace together with its tables. Until now, the code responsible for preparing mutations for a new table could assume the existence of the keyspace. This assumption wasn't necessary, but removing it required some refactoring.

Note that Alternator already had exactly the same need - creating a keyspace together with a table - and I think solved it correctly without needing any refactoring (but if you think there was anything wrong with that code, please let me know). The Alternator code is in alternator/executor.cc, create_table_on_shard0(), and had the following code near the end of the function:
```c++
    auto group0_guard = co_await mm.start_group0_operation();
    schema_mutations = co_await create_keyspace(keyspace_name, sp, gossiper,
    db::schema_tables::add_table_or_view_to_schema_mutation(schema, ts, true, schema_mutations);
    co_await seastar::async([&] {
        mm.get_notifier().before_create_column_family(*schema, schema_mutations, ts);
    });
    co_await mm.announce(std::move(schema_mutations), std::move(group0_guard), format("alternator-executor: create {} table", table_name));
```
the actual code has some comments explaining why I had to call these bizarre functions.
    

> 
> Fixes #15437

File: table_helper.cc
Comment (nyh) [inline table_helper.cc:132]: You had some nice suggestions on how to change things more radically (e.g., drop this table_helper) in the commit message. It's a shame nobody will see it later. Maybe you should copy some of them into this comment.

**Review by nyh — CHANGES_REQUESTED** [review]
Looks good, especially the redis patch would (hopefully...) noticeably improve things :-) I just left a few minor comments and requests. Thanks!

**Comment by patjed41** [comment]
> Note that Alternator already had exactly the same need - creating a keyspace together with a table - and I think solved it correctly without needing any refactoring (but if you think there was anything wrong with that code, please let me know). The Alternator code is in alternator/executor.cc, create_table_on_shard0(), and had the following code near the end of the function:
> 
> ```c++
>     auto group0_guard = co_await mm.start_group0_operation();
>     schema_mutations = co_await create_keyspace(keyspace_name, sp, gossiper,
>     db::schema_tables::add_table_or_view_to_schema_mutation(schema, ts, true, schema_mutations);
>     co_await seastar::async([&] {
>         mm.get_notifier().before_create_column_family(*schema, schema_mutations, ts);
>     });
>     co_await mm.announce(std::move(schema_mutations), std::move(group0_guard), format("alternator-executor: create {} table", table_name));
> ```

I suspect that the Alternator code is wrong.

Two problems require refactoring:
1. Calling `db.find_keyspace(cfm->ks_name())` in `migration_manager::prepare_new_column_family_announcement` when the keyspace doesn't exist.
2. Calling `_db.find_keyspace(s.ks_name())` in `tablet_allocator_impl::on_before_create_column_family` when the keyspace doesn't exist.

If I understand the code correctly, Alternator solved the first problem differently, but it didn't solve the second problem (it didn't exist then). All Alternator tests pass because they are run with tablets turned off. When I turn them on, they start failing.

I think this problem has to be solved sooner or later. My PR solves it by passing `keyspace_metadata` to `on_before_create_column_family`. So, regarding @tgrabiec's comment, since Alternator also needs these changes, I don't think we cannot avoid them.

Well, we don't have to implement the new `prepare_new_column_family_announcement` function as I did, and we can use the Alternator's approach instead, but there are so many places with the same problem that I think introducing a new function is a good idea. I can also refactor `create_table_on_shard0` to use the new function. I didn't notice it earlier.

File: alternator/executor.cc
Comment (patjed41) [reply alternator/executor.cc:1123]: > If you had wanted to use the new "ksm" variable somewhere, I could understand this split, but it doesn't seem that you actually use this variable anywhere.

I do it in the "migration_listener: add parameter to on_before_create_column_family" commit. That is why I've made this refactor.

File: redis/keyspace_utils.cc
Comment (patjed41) [reply redis/keyspace_utils.cc:194]: I could add it, but I'm fixing it in the follow-up -- #15613, so I don't think it's necessary.

File: redis/keyspace_utils.cc
Comment (patjed41) [reply redis/keyspace_utils.cc:203]: I think we've already been creating all tables in one go. 63d3449bc3e0434e2c77e470477d83d882ee37ca reduced the number of `announce`s to two -- one for all keyspaces and one for all tables. This PR only merges these two `announce`s into one. So, I guess that this issue was simply forgotten and we can close it.

File: table_helper.cc
Comment (patjed41) [reply table_helper.cc:132]: Sure.

Can I consider this suggestion to be a general way to go in this kind of situation? I saw FIXMEs with a one-line comment and a lot of information (like suggestions on how to fix them) in the commit message, so I'm a bit confused.

File: service/migration_manager.cc
Comment (patjed41) [reply service/migration_manager.cc:720]: Good catch, I'll fix it.

File: redis/keyspace_utils.cc
Comment (patjed41) [reply redis/keyspace_utils.cc:203]: Well, I haven't checked it yet, so another option is that this issue still exists, but for a different reason than the number of `announce`s.

Comment (nyh) [reply redis/keyspace_utils.cc:203]: Interesting, I updated #5328 with this information, and we should check it and close it.

File: table_helper.cc
Comment (nyh) [reply table_helper.cc:132]: Sometimes the way to fix something is fairly obvious, we just need to remember to do it, so the FIXME doesn't need to be long. Here what bothered me wasn't the shortness of the FIXME per se, but rather the fact that the FIXME suggests we should do one thing (call announce once) whereas the commit message suggests a completely different approach (removing this function altogether, not fixing it). So you should decide which fix you're proposing, and if it's the one in the commit message - the FIXME shouldn't be this line you added here but a FIXME in another place saying "don't use table_helper, use X and Y instead".

**Comment by nyh** [comment]
> > Note that Alternator already had exactly the same need - creating a keyspace together with a table - and I think solved it correctly without needing any refactoring (but if you think there was anything wrong with that code, please let me know). The Alternator code is in alternator/executor.cc, create_table_on_shard0(), and had the following code near the end of the function:
> > ```c++
> >     auto group0_guard = co_await mm.start_group0_operation();
> >     schema_mutations = co_await create_keyspace(keyspace_name, sp, gossiper,
> >     db::schema_tables::add_table_or_view_to_schema_mutation(schema, ts, true, schema_mutations);
> >     co_await seastar::async([&] {
> >         mm.get_notifier().before_create_column_family(*schema, schema_mutations, ts);
> >     });
> >     co_await mm.announce(std::move(schema_mutations), std::move(group0_guard), format("alternator-executor: create {} table", table_name));
> > ```
> 
> I suspect that the Alternator code is wrong.
> 
> Two problems in this PR require refactoring:
> 
>     1. Calling `db.find_keyspace(cfm->ks_name())` in `migration_manager::prepare_new_column_family_announcement` when the keyspace doesn't exist.
> 
>     2. Calling `_db.find_keyspace(s.ks_name())` in `tablet_allocator_impl::on_before_create_column_family` when the keyspace doesn't exist.
> 
> 
> If I understand the code correctly, Alternator solved the first problem differently,

Exactly.

> but it didn't solve the second problem (it didn't exist then). All Alternator tests pass because they are run with tablets turned off. When I turn them on, they start failing.

Oops. That's a Tablet bug - that changed the API (or the assumption of the API, namely that on_before_create_column_family may lookup the keyspace) and broke Alternator, so please fix Alternator as well. You can change Alternator to call your new functions - I didn't like the functions I was using (it was really ugly calling the notifiers manually) but didn't have a choice at the time - if you are creating new, better, functions please change Alternator to use them as well (should be easy).

> I think this problem has to be solved sooner or later. My PR solves it by passing `keyspace_metadata` to `on_before_create_column_family`. So, regarding @tgrabiec's comment, since Alternator also needs these changes, I don't think we cannot avoid them.

I didn't see @tgrabiec's comment about Alternator, but I agree.
 
> Well, we don't have to implement the new `prepare_new_column_family_announcement` function as I did, and we can use the Alternator's approach instead, but there are so many places with the same problem that I think introducing a new function is a good idea.

I agree.

> I can also refactor `create_table_on_shard0` to use the new function. I didn't notice it earlier.

Thanks. I don't think you need much refactoring - there are three or so consecutive lines that basically map to the same thing that your new function is doing, so you can just call it instead of these three lines.

File: service/tablet_allocator.cc
Comment (patjed41) [reply service/tablet_allocator.cc:828]: > The question is, in the case of keyspace already existing, whether this `create_replication_strategy` will return the same thing as `ks.get_replication_strategy()` would. If it turns out these replication strategies are different in a way that may affect the behavior of below code, this is unsafe.

Let's assume that the keyspace exists.

If the new `prepare_new_column_family_announcement` is called with `keyspace_metadata` different than the actual metadata, then of course these strategies can be different. I don't know why someone would do this though.

If the new `prepare_new_column_family_announcement` is called with actual `keyspace_metadata`, I'm not sure if these strategies can be different. The question is whether a keyspace's replication strategy always stays consistent with its metadata. In other words, whether `ks.get_replication_strategy()` and `abstract_replication_strategy::create_replication_strategy(ks.metadata()->strategy_name(), ks.metadata()->strategy_options())` always describe identical strategies. I think it's true, but unfortunately I couldn't find a proof.

If the above is true, we can solve the first case by prohibiting a call to the new `prepare_new_column_family_announcement` if the keyspace exists (for example, by throwing an exception).

Otherwise, we cannot use the `keyspace_metadata` parameter in `tablet_allocator_impl::on_before_create_column_family` if the keyspace exists even if it came from the old `prepare_new_column_family_announcement`. The only solution I see is using the `keyspace_metadata` parameter there only if the keyspace doesn't exist. I like this solution less because it is less general and doesn't prevent similar problems from appearing in other places.

> Even if the keyspace does not exist yet, it could be that creating it first and then calling `get_replication_strategy()` would return a different thing than doing `create_replication_strategy` directly here, again causing different behavior of below code. (I don't believe that this happens though, after all, whatever `get_replication_strategy()` returns must have been created somehow, we just need to find a way to recreate the same thing. But you should check.)

Let's assume that the keyspace doesn't exist.

`prepare_new_keyspace_announcement` calls `make_create_keyspace_mutations` that creates mutations for the strategy options and the strategy name using `keyspace_metadata` in exactly the opposite way to how `create_keyspace_from_schema_partition` creates `keyspace_metadata`. In `database::parse_system_tables`, `keyspace_metadata` received from `create_keyspace_from_schema_partition` is passed unchanged to `create_keyspace` and to `create_in_memory_keyspace` that calls `keyspace::create_replication_strategy`, which uses the strategy name and the strategy options from the unchanged metadata to create the replication strategy.

So, it looks like calling `get_replication_strategy()` after creating the keyspace and calling `create_replication_strategy` directly would return the same strategy.

Comment (patjed41) [reply service/tablet_allocator.cc:828]: After discussing it with @kbr-scylla, we concluded that:
- we can safely use `keyspace_metadata` and the `abstract_replication_strategy::create_replication_strategy` parameter in `tablet_allocator_impl::on_before_create_column_family` assuming the keyspace doesn't exist or it exists and `keyspace_metadata` is actual,
- in `tablet_allocator_impl::on_before_create_column_family`, calling `ks.get_replication_strategy()` after creating the keyspace and calling `abstract_replication_strategy::create_replication_strategy` directly when the keyspace doesn't exist returns the same strategy.

So, we only have to worry about this:
> Let's assume that the keyspace exists.
>
> If the new `prepare_new_column_family_announcement` is called with `keyspace_metadata` different than the actual metadata, then of course these strategies can be different. I don't know why someone would do this though.

We decided that a simple sanity check, that ensures the new `prepare_new_column_family_announcement` is called only when the keyspace doesn't exist, is enough. After all, this function has been introduced only to handle a case when the keyspace doesn't exist.

File: alternator/executor.cc
Comment (nyh) [inline alternator/executor.cc:1134]: I see you had to use this back_inserter idiom in several places, which suggests that maybe the `prepare_new_column_family_annoucement()` function, instead of returning a new vector, should append mutations to a given vector? It will make the different uses simpler?

Just like the old function "add_table_or_view_to_schema_mutation" which added something to an existing vector, and so do the notifiers.

**Review by nyh — APPROVED** [review]

File: alternator/executor.cc
Comment (patjed41) [reply alternator/executor.cc:1134]: That's a good idea, I'll do it. After all, the new `prepare_new_column_family_annoucement` has been introduced to be called together with `prepare_new_keyspace_announcement`. So, it makes sense for this function to behave differently from the other `prepare_` functions that usually return all mutations for an announcement.

**Comment by patjed41** [comment]
v4: Changed the new `prepare_new_keyspace_announcement` so that it updates a vector of mutation by reference.

File: service/tablet_allocator.cc
Comment (gleb-cloudius) [inline service/tablet_allocator.cc]: Shouldn't we assume existence of the keyspace any longer?

Comment (gleb-cloudius) [reply service/tablet_allocator.cc]: > Shouldn't we assume existence of the keyspace any longer?

OK. In the next patch.

File: table_helper.cc
Comment (gleb-cloudius) [inline table_helper.cc:124]: > Another approach is removing table_helper completely.

+1. It was added in hope that it will be reused for many tables, but since it was not there is not point in having it.

**Review by gleb-cloudius — APPROVED** [review]

**Comment by patjed41** [comment]
rebased on the current master with more stable CI

**Comment by patjed41** [comment]
@nyh ping

**Comment by patjed41** [comment]
@nyh This PR is ready to be merged, and it's blocking my progress with concurrent bootstrap.

**Comment by nyh** [comment]
> @nyh please merge

Sorry, I'll merge now.

What happens after this PR with Alternator + Tablets? Did you fix it, or it remains broken after this PR? BTW, when you guys test tablets, don't you run the entire test suite with tablets enabled by default?

**Comment by kbr-scylla** [comment]
> BTW, when you guys test tablets, don't you run the entire test suite with tablets enabled by default?

No, a subset of suites runs with tablets, a subset runs without. Apparently the alternator suite runs without tablets.

**Comment by patjed41** [comment]
> What happens after this PR with Alternator + Tablets? Did you fix it, or it remains broken after this PR?

This PR fixes the Alternator + Tablets issue.

**Comment by avikivity** [comment]
I reverted it due to the coroutine::parallel_for_each bug (scylladb/seastar#1902). Please rebase and resubmit once the seastar submodule is updated.

**Comment by avikivity** [comment]
The revert: d450a145ce16660d37b02c2db421b944cb973921

**Comment by kbr-scylla** [comment]
Damn I thought I dequeued it.
Ah, maybe I forgot to cancel a concurrently running promotion which unluckily succeeded that time.

**Comment by kbr-scylla** [comment]
I dequeued this one: https://github.com/scylladb/scylladb/pull/15613

Didn't think that this PR was also affected :/

**Pull request merged**
